## Dependencies 

In [4]:
## Packages : 
%pip install scikit-learn

Note: you may need to restart the kernel to use updated packages.


In [3]:
import numpy as np
import pandas as pd
## to split the data
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
## to evaluate the model 
from sklearn.metrics import accuracy_score

# Upload the data 

In [10]:
# Load the CSV file 
heart_db = pd.read_csv("../data/heart_disease_data.csv")

# first 5 rows of the db
heart_db.head(5)

age  sex  cp  trestbps  chol  fbs  restecg  thalach  exang  oldpeak  slope  \
0   63    1   3       145   233    1        0      150      0      2.3      0   
1   37    1   2       130   250    0        1      187      0      3.5      0   
2   41    0   1       130   204    0        0      172      0      1.4      2   
3   56    1   1       120   236    0        1      178      0      0.8      2   
4   57    0   0       120   354    0        1      163      1      0.6      2   

   ca  thal  target  
0   0     1       1  
1   0     2       1  
2   0     2       1  
3   0     2       1  
4   0     2       1

In [12]:
heart_db['trestbps'].unique()

array([145, 130, 120, 140, 172, 150, 110, 135, 160, 105, 125, 142, 155,
       104, 138, 128, 108, 134, 122, 115, 118, 100, 124,  94, 112, 102,
       152, 101, 132, 148, 178, 129, 180, 136, 126, 106, 156, 170, 146,
       117, 200, 165, 174, 192, 144, 123, 154, 114, 164])

# Data analyzing and processing

In [4]:
# size of db
heart_db.shape

(303, 14)

=> The data base is small

In [5]:
# more informations of data
heart_db.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 303 entries, 0 to 302
Data columns (total 14 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       303 non-null    int64  
 1   sex       303 non-null    int64  
 2   cp        303 non-null    int64  
 3   trestbps  303 non-null    int64  
 4   chol      303 non-null    int64  
 5   fbs       303 non-null    int64  
 6   restecg   303 non-null    int64  
 7   thalach   303 non-null    int64  
 8   exang     303 non-null    int64  
 9   oldpeak   303 non-null    float64
 10  slope     303 non-null    int64  
 11  ca        303 non-null    int64  
 12  thal      303 non-null    int64  
 13  target    303 non-null    int64  
dtypes: float64(1), int64(13)
memory usage: 33.3 KB


=> All the columns has no missing values ( = 303 non_null ) 

In [6]:
# cheking for messing values 
heart_db.isnull().sum()

age         0
sex         0
cp          0
trestbps    0
chol        0
fbs         0
restecg     0
thalach     0
exang       0
oldpeak     0
slope       0
ca          0
thal        0
target      0
dtype: int64

In [7]:
# statistical measures about the data 
heart_db.describe()

age         sex          cp    trestbps        chol         fbs  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean    54.366337    0.683168    0.966997  131.623762  246.264026    0.148515   
std      9.082101    0.466011    1.032052   17.538143   51.830751    0.356198   
min     29.000000    0.000000    0.000000   94.000000  126.000000    0.000000   
25%     47.500000    0.000000    0.000000  120.000000  211.000000    0.000000   
50%     55.000000    1.000000    1.000000  130.000000  240.000000    0.000000   
75%     61.000000    1.000000    2.000000  140.000000  274.500000    0.000000   
max     77.000000    1.000000    3.000000  200.000000  564.000000    1.000000   

          restecg     thalach       exang     oldpeak       slope          ca  \
count  303.000000  303.000000  303.000000  303.000000  303.000000  303.000000   
mean     0.528053  149.646865    0.326733    1.039604    1.399340    0.729373   
std      0.525860   22.905161    0.469794    1.161075    0.616226    1.022606   
min      0.000000   71.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.000000  133.500000    0.000000    0.000000    1.000000    0.000000   
50%      1.000000  153.000000    0.000000    0.800000    1.000000    0.000000   
75%      1.000000  166.000000    1.000000    1.600000    2.000000    1.000000   
max      2.000000  202.000000    1.000000    6.200000    2.000000    4.000000   

             thal      target  
count  303.000000  303.000000  
mean     2.313531    0.544554  
std      0.612277    0.498835  
min      0.000000    0.000000  
25%      2.000000    0.000000  
50%      2.000000    1.000000  
75%      3.000000    1.000000  
max      3.000000    1.000000

In [8]:
# checking the distribution of target variable 
heart_db['target'].value_counts()

1    165
0    138
Name: target, dtype: int64

1 : Defective heart

2: Healthy heart 


--> In this case, the two classes are almost equal ,so there is no problem with data imbalance

In [9]:
# Splitting the features and target 
X = heart_db.drop(columns='target' , axis =1)
Y = heart_db['target']


## Splitting the data intro ( training and test data )

Test data : 20%

Training data : 80%

 stratify =Y : The target(Y) is balanced between the two data ( test and train ) =>  the test and train has almost the same number of 0 and 1

random_state : The way of splitting the data ( it's changeable )

In [10]:
X_train  , X_test , Y_train  , Y_test = train_test_split(X , Y , test_size=0.2 , stratify =Y , random_state =2)

In [11]:
print(X.shape , X_train.shape , X_test.shape)

(303, 13) (242, 13) (61, 13)


whole data : 303

Train data (80%) : 242

Test data : (20%) : 61

# Model (Logistic Regression )

## Train the model 

In [12]:
# load the model
model = LogisticRegression()

In [13]:
# train the Logistic model
model.fit(X_train ,Y_train)

/home/nourguealia/anaconda3/lib/python3.9/site-packages/sklearn/linear_model/_logistic.py:814: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

##  Evaluate the model

###  Evaluate with Training data

In [14]:
# Evaluation the model : Training data
Y_training_predicted = model.predict(X_train)
Training_accuracy_training = accuracy_score(Y_training_predicted , Y_train)

In [15]:
print("The accuracy of the training data = ", Training_accuracy_training)

The accuracy of the training data =  0.8512396694214877


###  Evaluate with Test data

In [16]:
# Evaluation the model : Test data
Y_test_predicted = model.predict(X_test)
Training_accuracy_test = accuracy_score(Y_test_predicted , Y_test)

In [17]:
print("The accuracy of the test data = ", Training_accuracy_test)

The accuracy of the test data =  0.819672131147541


# Save the model

In [19]:
import pickle 

### Save

In [21]:
## name of file saved
filename = 'model.sav'
## save the 'model' with the nale 'model.sav' with a WB option => w:write , b:binary
pickle.dump(model , open(filename ,'wb'))

### Load

In [23]:
## loading the saved model , with RB option => r:reading , b:binary
loaded_model = pickle.load(open('model.sav' ,'rb'))